firebase-vscode/src/workflow.ts

```diff
@@ -0,0 +1,429 @@
+  const rootFolders = getRootFolders();
+  const filePath = path.join(rootFolders[0], 'firebase-plugin-debug.log');
+  pluginLogger.info('Logging to path', filePath);
+  if (!process.env.MONOSPACE_ENV) {
+    logger.add(
+      new transports.File({
+        level: "debug",
+        filename: filePath,
+        format: format.printf((info) => {
+          const segments = [info.message, ...(info[SPLAT] || [])]
+            .map(tryStringify);
+          return `[${info.level}] ${stripAnsi(segments.join(" "))}`;
+        }),
+      })
+    );
+  }
+  // Read config files and store in memory.
+  readFirebaseConfigs();
+  // Check current users state
+  fetchUsers();
+  fetchChannels();
+
+  /**
+   * Call pluginLogger with log arguments received from webview.
+   */
+  broker.on("writeLog", async ({ level, args }) => {
+    pluginLogger[level]('(Webview)', ...args);
+  });
+
+  broker.on("getEnv", async () => {
+    pluginLogger.debug(`Value of process.env.MONOSPACE_ENV: `
+      + `${process.env.MONOSPACE_ENV}`);
+    broker.send("notifyEnv", {
+      env: {
+        isMonospace: Boolean(process.env.MONOSPACE_ENV),
+      }
+    });
+  });
+
+  broker.on("getUsers", async () => {
+    if (users.length === 0) {
+      await fetchUsers();
+    }
+    broker.send("notifyUsers", { users });
+    currentUserEmail = updateCurrentUser(users, broker);
+  });
+
+  broker.on("logout", async ({ email }: { email: string }) => {
+    try {
+      await logoutUser(email);
+      const accounts = await getAccounts();
+      users = accounts.map((account) => account.user);
+      broker.send("notifyUsers", { users });
+      currentUserEmail = updateCurrentUser(users, broker);
+    } catch (e) {
+      // ignored
+    }
+  });
+
+  broker.on("getSelectedProject", async () => {
+    // For now, just read the cached value.
+    // TODO: Extend this to reading from firebaserc
+    if (firebaseRC?.projects?.default) {
+      broker.send("notifyProjectChanged",
+        { projectId: firebaseRC?.projects?.default });
+    }
+    fetchChannels();
+  });
+
+  broker.on("showMessage", async ({ msg, options }) => {
+    vscode.window.showInformationMessage(msg, options);
+  });
+
+  broker.on("openLink", async ({ href }) => {
+    vscode.env.openExternal(vscode.Uri.parse(href));
+  });
+
+  broker.on("addUser", async () => {
+    const { user } = await login();
+    users.push(user);
+    if (users) {
+      broker.send("notifyUsers", { users });
+      currentUserEmail = updateCurrentUser(
+        users,
+        broker,
+        user.email
+      );
+    }
+  });
+
+  broker.on("requestChangeUser", (
+    { user: requestedUser }:
+      { user: User | ServiceAccountUser }
+  ) => {
+    if (users.some((user) => user.email === requestedUser.email)) {
+      currentUserEmail = requestedUser.email;
+      broker.send("notifyUserChanged", { email: currentUserEmail });
+    }
+  });
+
+  broker.on("selectProject", async ({ email }) => {
+    let projectId;
```

Comment:
Optional: Some of these broker.on calls have a lot of logic and functionality tucked away in anonymous functions. It may be worth separating them out into named functions (or function generators) for cleaner organization and easier unit testing.

Reply:
Separated out the 2 largest ones.